[Publishing] Add Publish to S3 using Buildkite

ParaskP7 opened the pull request:
App Infra Request: pdnsEh-1RV-p2

Depends On:
- [x] Security related inquiry/discussion on p1727693975599849-slack-C02KLTL3MKM.
- [x] [BuildkiteCI#526](https://github.com/Automattic/buildkite-ci/pull/526)
- [x] Configure `.mobile-secrets` (create deployment key & environmental variables),  and `GitHub` (add deployment key).

### Description

This PR adds publishing to our S3 Maven using Buildkite:

### Testing Instructions

`TODO`:
- Wait for this [BuildkiteCI#526](https://github.com/Automattic/buildkite-ci/pull/526) PR to get deployed/merged.
- Trigger a CI build on this PR and wait for the library to be published on our S3 Maven.
- Use a project like [Gravatar-SDK-Android](https://github.com/Automattic/Gravatar-SDK-Android) to update the group from `com.github.yalantis` to `com.automattic`, while using a `<PR>-<COMMIT_SHA>` combination as its test version.
- Build the test project and verify that any `uCrop` related functionality is still working as expected.

ParaskP7 committed:
Build: Update gradle distribution to 7.5

Command: ./gradlew wrapper --gradle-version=7.5 --distribution-type=all

Release Notes: https://docs.gradle.org/7.5/release-notes.html
Distribution: https://services.gradle.org/distributions/
gradle-7.5.1-all.zip

------------------------------------------------------------------------

Note, although the 'distributionUrl' was already pointing to
'gradle-7.5-all.zip', the 'jar' and 'gradlew' files weren't up-to-date.
This was probably because when the previous update happened, the update
was either done manually by just updating the version on that url, or
the command was run once, and not twice, which is when it actually
works and updates all relevant Gradle related files.

ParaskP7 committed:
Build: Migrate to gradle plugin dsl

ParaskP7 committed:
Build: Remove all publishing related configuration

ParaskP7 committed:
Build: Add support for publishing the library to s3

ParaskP7 committed:
Docs: Update using and publishing section of readme

ParaskP7 committed:
CI: Add buildkite pipeline

AdamGrzybkowski commented on README.md:
```diff
@@ -195,6 +186,14 @@ We’d be really happy if you sent us links to your projects where you use our c
+In the following cases, the CI will publish a new version with the following format to our remote Maven repo:
+
+* For each commit in an open PR: `<PR-number>-<commit full SHA1>`
+* Each time a PR is merged to `trunk`: `trunk-<commit full SHA1>`
```
Are we renaming `develop` to `trunk`? I probably forgot to add one piece of information in the request - this repo has two important branches `develop` and `develop-non-native`. The one that we're interested in and planning to use is the second non-native one.

ParaskP7 replied:
Ah apologies @AdamGrzybkowski , this is a copy-pasting mistake, which I took from similar instructions we have on every other library of ours, I'll go ahead and rename `trunk` to `develop-non-native` on this `README.md` file.

> I probably forgot to add one piece of information in the request - this repo has two important branches develop and develop-non-native. The one that we're interested in and planning to use is the second non-native one.

Yea, I noticed that when I cloned the repo, seeing `develop-non-native` being the default branch, and so, I based everything on that, while ignoring `develop` for now. Thanks for the clarification! 🙇

ParaskP7 replied:
Also, now that we're talking about this, maybe I should delete this line as well, wdyt?

```
``` implementation 'com.automattic:ucrop:2.2.10-native' ``` - get power of the native code to preserve image quality (+ about 1.5 MB to an apk size)
```

And only keep this one line there:

```
``` implementation 'com.automattic:ucrop:2.2.10' ``` - lightweight general solution
```

Just to make it clear that we are only dealing/publishing the `non-native` side of things. 🤔

AdamGrzybkowski replied:
Sounds good to me 👍

ParaskP7 replied:
This is now done: 404636a

ParaskP7 committed:
Docs: Fix and improve the using and publishing section of readme

1. Remove reference to 'com.automattic:ucrop:2.2.10-native'.
2. Replace 'trunk' with 'develop-non-native'.

PR Comment: https://github.com/Automattic/uCrop/pull/2
#discussion_r1782277198

wzieba reviewed (commented):
Looks good to me! Let's see results when CI will be triggered

wzieba commented on README.md:
````diff
@@ -15,18 +15,7 @@
-    ``` implementation 'com.github.yalantis:ucrop:2.2.9-native' ``` - get power of the native code to preserve image quality (+ about 1.5 MB to an apk size)
+    ``` implementation 'com.automattic:ucrop:2.2.10' ``` - lightweight general solution
 
 2. Add UCropActivity into your AndroidManifest.xml
````
It's not a scope of this PR, but I think that the SDK should add this activity to its own manifest and rely on manifest merge to include it into client's manifest.

ParaskP7 committed:
Merge branch 'develop-non-native' of github.com:Automattic/uCrop into publishing/add-publish-to-s3-using-buildkite

ParaskP7 committed:
Build: Set org gradle jvmargs to fix an out of memory error on ci

While running the 'Publish to S3' step on CI the build fails with the
below exception:

"Exception in thread "DefaultDispatcher-worker-2"
java.lang.OutOfMemoryError: Metaspace
Daemon will be stopped at the end of the build after running out of
JVM memory"

This error occurs when the Metaspace region of the JVM's memory on CI is
exhausted. As such, the default Metaspace size may be too small for the
CI build process. Increasing the Metaspace size by adding the JVM
arguments will hopefully resolve this problem.

ParaskP7 commented:
**UPDATE**:

1. After configuring `.mobile-secrets` and GitHub, I run this [first](https://buildkite.com/automattic/ucrop/builds/1) build, which verified that the CI setup has been successful (🎉) but nevertheless `Publish to S3` failed with an out of memory exception. ❌
2. After updating `org.gradle.jvmargs` (see f881f29) the [second](https://buildkite.com/automattic/ucrop/builds/2#0192486a-0c78-4b5a-abed-bd4b343308f8) manual build ended-up in a total success (🎉) and our own version of `uCrop` got published to `s3` for the first time: ✅ 

```
petros.paraskevopoulos@Mac-Studio uCrop % aws s3 ls a8c-libs/android/com/automattic/ucrop/
                           PRE publishing_add-publish-to-s3-using-buildkite-f881f29c6a1802963b98c68f44b0b840b2417e4d/
2024-10-01 17:17:04        566 maven-metadata.xml
2024-10-01 17:17:04         32 maven-metadata.xml.md5
2024-10-01 17:17:04         40 maven-metadata.xml.sha1
2024-10-01 17:17:04         64 maven-metadata.xml.sha256
2024-10-01 17:17:04        128 maven-metadata.xml.sha512
```

3. I'll now try and see if I can use this new `<PR>-<COMMIT_SHA>` combination on the [Gravatar-SDK-Android](https://github.com/Automattic/Gravatar-SDK-Android) repo and that it compiles/works as expected. 🤞

ParaskP7 commented:
**UPDATE** on [(3.)](https://github.com/Automattic/uCrop/pull/2#issuecomment-2386121149) above:

I tested our version of `uCrop`, the `Automattic/uCrop`, on [Gravatar-SDK-Android](https://github.com/Automattic/Gravatar-SDK-Android), using a `<BRANCH>-<COMMIT_SHA>` instead (the `<PR>-<COMMIT_SHA>` combination wasn't available, it probably being a manual, and not a PR build).

Result: Everything is working as expected! ✅ 

Below is a patch for you to try it yourself @AdamGrzybkowski, just to verify that everything is working as expected with this version of `uCrop`. Then, we could merge this and create a new release, as per the `README.md`, the `2.2.10`, or else, the `com.automattic:ucrop:2.2.10`, and start using it:

<details>
    <summary>Batch</summary>

```diff
Index: gradle/libs.versions.toml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/gradle/libs.versions.toml b/gradle/libs.versions.toml
--- a/gradle/libs.versions.toml	(revision 345e88842413555037409b599a5d9aaec2081ed4)
+++ b/gradle/libs.versions.toml	(date 1727793809388)
@@ -15,7 +15,7 @@
 mockk = "1.13.9"
 appcompat = "1.7.0"
 material = "1.12.0"
-ucrop = "2.2.9"
+ucrop = "publishing_add-publish-to-s3-using-buildkite-f881f29c6a1802963b98c68f44b0b840b2417e4d"
 coil = "2.7.0"
 activityCompose = "1.9.2"
 robolectric = "4.12.2"
@@ -64,7 +64,7 @@
 retrofit = { group = "com.squareup.retrofit2", name = "retrofit", version.ref = "retrofit" }
 retrofit-moshi-converter = { group = "com.squareup.retrofit2", name = "converter-moshi", version.ref = "retrofit" }
 okhttp = { group = "com.squareup.okhttp3", name = "okhttp", version.ref = "okhttp" }
-ucrop = { group = "com.github.yalantis", name = "ucrop", version.ref = "ucrop" }
+ucrop = { group = "com.automattic", name = "ucrop", version.ref = "ucrop" }
 coil-compose = { group = "io.coil-kt", name = "coil-compose", version.ref = "coil" }
 coil-svg = { group = "io.coil-kt", name = "coil-svg", version.ref = "coil" }
 coil-test = { group = "io.coil-kt", name = "coil-test", version.ref = "coil" }
Index: settings.gradle.kts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/settings.gradle.kts b/settings.gradle.kts
--- a/settings.gradle.kts	(revision 345e88842413555037409b599a5d9aaec2081ed4)
+++ b/settings.gradle.kts	(date 1727793854753)
@@ -26,6 +26,13 @@
         google()
         mavenCentral()
         maven { url = uri("https://jitpack.io") }
+        maven {
+            url = uri("https://a8c-libs.s3.amazonaws.com/android")
+            content {
+                includeGroup("com.automattic")
+                includeGroup("com.automattic.ucrop")
+            }
+        }
     }
 }
```

</details>

PS: Also, from my screenshot below you will notice that we are indeed bringing the correct version of `BitmapCropTask`, the one with the changes you need:

![image](https://github.com/user-attachments/assets/a2fa9c2f-4680-4a9a-9b70-787abfdee062)

Cc @wzieba

AdamGrzybkowski commented:
>Below is a patch for you to try it yourself @AdamGrzybkowski, just to verify that everything is working as expected with this version of uCrop.

Thanks for the patch - I tested it and all worked fine!

ParaskP7 commented:
Awesome, thanks for testing @AdamGrzybkowski , I'll proceed with merging this soon and then will try publishing our new `om.automattic:ucrop:2.2.10` release, for it to then be uploaded to S3, hoping that all the tooling work I did so far would suffice. 🤞
